src/gui/fspathedit.cpp

```diff
@@ -0,0 +1,361 @@
+    : q_ptr {q}
+    , m_editor {editor}
+    , m_browseAction {new QAction(q)}
+    , m_browseBtn {new QToolButton(q)}
+    , m_mode {FileSystemPathEdit::Mode::FileOpen}
+{
+    m_browseAction->setIconText(browseButtonBriefText.tr());
+    m_browseAction->setText(browseButtonFullText.tr());
+    m_browseAction->setToolTip(browseButtonFullText.tr().remove(QLatin1Char('&')));
+    m_browseAction->setShortcut({Qt::CTRL + Qt::Key_B});
+    m_browseBtn->setDefaultAction(m_browseAction);
+    m_fileNameFilter = tr("Any file (*)");
+    m_editor->setBrowseAction(m_browseAction);
+    modeChanged();
+}
+
+void FileSystemPathEdit::FileSystemPathEditPrivate::browseActionTriggered()
+{
+    QString selectedPath;
+    Q_Q(FileSystemPathEdit);
+    QString filter = q->fileNameFilter();
+    QString directory = q->currentDirectory().isEmpty() ? QDir::homePath() : q->currentDirectory();
+
+    switch (m_mode) {
+    case FileSystemPathEdit::Mode::FileOpen:
+        selectedPath = QFileDialog::getOpenFileName(q, dialogCaptionOrDefault(), directory, filter);
+        break;
+    case FileSystemPathEdit::Mode::FileSave:
+        selectedPath = QFileDialog::getSaveFileName(q, dialogCaptionOrDefault(), directory, filter, &filter);
+        break;
+    case FileSystemPathEdit::Mode::DirectoryOpen:
+    case FileSystemPathEdit::Mode::DirectorySave:
+        selectedPath = QFileDialog::getExistingDirectory(q, dialogCaptionOrDefault(),
+                                directory, QFileDialog::DontResolveSymlinks | QFileDialog::ShowDirsOnly);
+        break;
+    default:
+        throw std::logic_error("Unknown FileSystemPathEdit mode");
+    }
+    if (!selectedPath.isEmpty())
+        q->setEditWidgetText(selectedPath);
+}
+
+QString FileSystemPathEdit::FileSystemPathEditPrivate::dialogCaptionOrDefault() const
+{
+    if (!m_dialogCaption.isEmpty())
+        return m_dialogCaption;
+
+    switch (m_mode) {
+    case FileSystemPathEdit::Mode::FileOpen:
+    case FileSystemPathEdit::Mode::FileSave:
+        return defaultDialogCaptionForFile.tr();
+    case FileSystemPathEdit::Mode::DirectoryOpen:
+    case FileSystemPathEdit::Mode::DirectorySave:
+        return defaultDialogCaptionForDirectory.tr();
+    default:
+        throw std::logic_error("Unexpected value of type FileSystemPathEdit::Mode");
+    }
+}
+
+void FileSystemPathEdit::FileSystemPathEditPrivate::modeChanged()
+{
+    QStyle::StandardPixmap pixmap = QStyle::SP_DialogOpenButton;
+    bool showDirsOnly = false;
+    switch (m_mode) {
+    case FileSystemPathEdit::Mode::FileOpen:
+    case FileSystemPathEdit::Mode::FileSave:
+        pixmap = QStyle::SP_DialogOpenButton;
+        showDirsOnly = false;
+        break;
+    case FileSystemPathEdit::Mode::DirectoryOpen:
+    case FileSystemPathEdit::Mode::DirectorySave:
+        pixmap = QStyle::SP_DirOpenIcon;
+        showDirsOnly = true;
+        break;
+    }
```

Comment:
I'm not sure if a default case would make sense... but it doesn't hurt to add one?